staging/src/k8s.io/apiserver/pkg/cel/library/ip_addrs.go

```diff
@@ -0,0 +1,151 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package library
+
+import (
+	"net/netip"
+
+	"github.com/google/cel-go/cel"
+	"github.com/google/cel-go/common/types"
+	"github.com/google/cel-go/common/types/ref"
+
+	apiservercel "k8s.io/apiserver/pkg/cel"
+)
+
+// IPAddr provides a CEL function library extension of IP address parsing functions.
+//
+// ipAddr
+//
+// Converts a string to an IP address or results in an error if the string is not a valid IP address.
+// The IP address must be an IPv4 or IPv6 address.
+//
+//	ipAddr(<string>) <IPAddr>
+//
+// Examples:
+//
+//	ipAddr('127.0.0.1') // returns an IPv4 address
+//	ipAddr('::1') // returns an IPv6 address
+//	ipAddr('127.0.0.256') // error
+//	ipAddr(':::1') // error
+//
+// isIPAddr
+//
+// Returns true if a string is a valid IP address. The IP address must be an IPv4 or IPv6 address.
+//
+//	isIPAddr(<string>) <bool>
+//
+// Examples:
+//
+//	isIPAddr('127.0.0.1') // returns true
+//	isIPAddr('::1') // returns true
+//	isIPAddr('127.0.0.256') // returns false
+//	isIPAddr(':::1') // returns false
+//
+// isIPv4 / isIPv6 / isInCIDR
+//
+// Returns true if a string is a valid IPv4 address, IPv6 address, or is in a CIDR range.
+//
+// - isIPv4: returns true when the IP Address is an IPv4 address.
+//
+// - isIPv6: returns true when the IP Address is an IPv6 address.
+//
+// Examples:
+//
+//	ipAddr('127.0.0.1').isIPv4Addr() // returns true
+//	ipAddr('::1').isIPv6Addr() // returns true
+//	ipAddr('127.0.0.1').isIPv6Addr() // returns false
+//	ipAddr('::1').isIPv4Addr() // returns false
+func IPAddrs() cel.EnvOption {
+	return cel.Lib(ipAddrsLib)
+}
+
+var ipAddrsLib = &ipAddrs{}
+
+type ipAddrs struct{}
+
+func (*ipAddrs) LibraryName() string {
+	return "k8s.ipaddrs"
+}
+
+var ipAddrLibraryDecls = map[string][]cel.FunctionOpt{
+	"ipAddr": {
+		cel.Overload("string_to_ipaddr", []*cel.Type{cel.StringType}, apiservercel.IPAddrType,
+			cel.UnaryBinding(stringToIpAddr))},
+	"isIPv4": {
+		cel.MemberOverload("ipaddr_is_ipv4", []*cel.Type{apiservercel.IPAddrType}, cel.BoolType,
+			cel.UnaryBinding(isIPv4))},
+	"isIPv6": {
+		cel.MemberOverload("ipaddr_is_ipv6", []*cel.Type{apiservercel.IPAddrType}, cel.BoolType,
+			cel.UnaryBinding(isIPv6))},
+	"isIPAddr": {
+		cel.Overload("is_ipaddr_string", []*cel.Type{cel.StringType}, cel.BoolType,
+			cel.UnaryBinding(isIPAddr))},
+}
+
+func (*ipAddrs) CompileOptions() []cel.EnvOption {
+	options := []cel.EnvOption{}
+	for name, overloads := range ipAddrLibraryDecls {
+		options = append(options, cel.Function(name, overloads...))
+	}
+	return options
+}
+
+func (*ipAddrs) ProgramOptions() []cel.ProgramOption {
+	return []cel.ProgramOption{}
+}
+
+func stringToIpAddr(arg ref.Val) ref.Val {
+	s, ok := arg.Value().(string)
+	if !ok {
+		return types.MaybeNoSuchOverloadErr(arg)
+	}
+
+	addr, err := netip.ParseAddr(s)
```

Comment:
So I've been pondering `netip` wrappers / better k8s IP utilities. https://github.com/danwinship/kubernetes/commits/ipfamily

My current attempt at a `ParseIP` function is [here](https://github.com/danwinship/kubernetes/commit/9bc20ff8cc4ff75d41b4e256889ba1b8c8134828#diff-311b2fd92f5d259a223ae13ba5792a4f9a7e50366815553214285b3924db31faR48-R61). As compared to this:

1. It returns an error if `addr.Zone() != ""`, which we definitely want here. (`netip.ParseAddr` allows strings such as `fe80::1234%eth0` which you need to be able to do sometimes, but which there are probably no good use cases for having in kubernetes API objects). (`net.ParseIP` does not allow zones in IPs, so this is the more-compatible-with-existing-code option anyway.)
2. It returns an error if `addr.Is4In6()`, i.e. `"::ffff:192.168.0.1"`, which again, there are some contexts where it's useful to be able to treat IPv4 addresses as IPv6 addresses, but kubernetes API objects are not such a context, so we should reject those too. (Most existing API objects *do* allow 4-in-6 addresses, and this is terrible, and there's probably at least one CVE lurking in there somewhere.)
3. It [overrides `netip.ParseAddr`'s error messages in some cases](https://github.com/danwinship/kubernetes/commit/9bc20ff8cc4ff75d41b4e256889ba1b8c8134828#diff-311b2fd92f5d259a223ae13ba5792a4f9a7e50366815553214285b3924db31faR114-R131), because `netip.ParseAddr`'s error messages are not especially helpful in the case where a user misunderstood what sort of data they were supposed to pass and passed something else (like an IP:port or a CIDR) instead. Actually, ParseAddr's error messages are pretty non-user-friendly in general and we might want to just return a static "invalid IP address" in any case that isn't one of the other mistakes.

(Assuming that branch or something like it eventually merges, then we can later make the functions here use the more generic functions there. But we should at least get the parsing right now.)